Client/Frontend/Browser/BrowserViewController.swift

```diff
@@ -1135,25 +1138,44 @@ class BrowserViewController: UIViewController {
         // Make the web view the first responder so that it can show the selection menu.
         return tabManager.selectedTab?.webView?.becomeFirstResponder() ?? false
     }
+
+    private func getCurrentAppState() -> AppState {
+        guard let homePanelController = self.homePanelController else {
+            if let tab = tabManager.selectedTab {
+                return .Tab(tabState: tab.tabState)
+            }
+            return .Loading
+        }
+        return .HomePanels(homePanelState: homePanelController.homePanelState)
+    }
+}
+
+extension BrowserViewController: AppStateDelegate {
+
+    func appDidUpdateState(appState: AppState) {
+        if AppConstants.MOZ_MENU {
+            menuViewController?.appState = appState
+        }
+    }
 }
 
 /**
  * History visit management.
  * TODO: this should be expanded to track various visit types; see Bug 1166084.
  */
 extension BrowserViewController {
-    func ignoreNavigationInTab(tab: Browser, navigation: WKNavigation) {
+    func ignoreNavigationInTab(tab: Tab, navigation: WKNavigation) {
```

Comment:
At what point do we make `BrowserViewController` a container for a `TabViewController` and a tool bar?